### PR TITLE
Providing input and output information per interaction

### PR DESCRIPTION
**Pull request by josojo**
This PR builds the logic to provide the input and output information for each interaction requested by the driver. This information is required for the driver to build the "local token conversation rule".

Testplan:
Unit tests only.

### CODE REVIEW

File: src/models/batch_auction_model.rs [hunk 1]
```diff
@@ -148,6 +148,28 @@ pub struct InteractionData {
...
+    /// `AMM -> GPv2Settlement`
+    #[serde(default)]
+    pub outputs: Vec<TokenAmount>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct TokenAmount {
+    #[serde(with = "u256_decimal")]
+    pub amount: U256,
+    pub token: H160,
```
Comment (josojo) [inline src/models/batch_auction_model.rs:172]: Copied from cowprotocol/service

**Review by bh2smith — APPROVED** [review]
Everything looks great here although I think I am missing some context (like the motivation for this change). From what I understand by reading the changes here, we will no longer insert the `exec_sell_amount` and `exec_by_amount` into the list of AMM objects that were provided with the batch? If that is the case, then I suppose this change will also be be needed in `solver-template-py`. I presume there will be some overlapping time for which both options are accepted as valid return data.

File: src/solve.rs [hunk 2]
```diff
@@ -224,6 +225,14 @@ fn build_payload_for_swap_and_approval(
             value: swap.value,
             call_data: swap.data.0.clone(),
             exec_plan: None,
+            inputs: vec![TokenAmount {
+                token: query.sell_token,
+                amount: swap.sell_amount,
```
Comment (josojo) [inline src/solve.rs:230]: Sometimes, for buy orders, the 0x api is buying too many tokens and then also receiving more tokens than what it got. But it seems that this is also not always happening. The logic will be a little more intrigued. I will do some more investigations and then prepare a solution in a separate PR. For sell-orders, this logic will always work.

**Comment by josojo** [comment]
>From what I understand by reading the changes here, we will no longer insert the exec_sell_amount and exec_by_amount into the list of AMM objects that were provided with the batch? 

The change that I am providing here is only for interaction data. I think the python solver does not yet deal with these, right? For interaction data the exec_sell_amount was not provided. Only for AMMs and orders, I think, but I am not 100% sure.

**Comment by fleupold** [comment]
I think what Ben is saying is that from the PR description alone it is not clear why this change is needed (I agree). It would be nice to include the extra context that for better slippage/internal buffer trading detection the result format is changing so that solvers need to explicitly state which tokens/amounts their interactions are intended to move.

Explicitly marking internal buffer trades as `"execution_plan": internal` will be a separate PR, right?

**Comment by josojo** [comment]
> It would be nice to include the extra context that for better slippage/internal buffer trading detection the result format is changing so that solvers need to explicitly state which tokens/amounts their interactions are intended to move.

okay

> Explicitly marking internal buffer trades as `"execution_plan": internal` will be a separate PR, right?

It is already merged here: https://github.com/gnosis/cow-dex-solver/pull/26

File: src/models/batch_auction_model.rs [hunk 3]
```diff
@@ -425,6 +447,59 @@ mod tests {
...
+                    token: H160([3; 20]),
+                    amount: 3000.into(),
+                },
+            ],
+        };
+        let expected_string = r#"{"target":"0xffffffffffffffffffffffffffffffffffffffff","value":"0x1","call_data":[1,2],"exec_plan":null,"inputs":[{"amount":"9999","token":"0x0101010101010101010101010101010101010101"}],"outputs":[{"amount":"2000","token":"0x0202020202020202020202020202020202020202"},{"amount":"3000","token":"0x0303030303030303030303030303030303030303"}]}"#;
+        assert_eq!(
+            serde_json::to_string(&interaction_data).unwrap(),
+            expected_string
+        );
```
Comment (fedgiac) [inline src/models/batch_auction_model.rs]: ~I didn't notice any difference with the block of code above, it appears to be unneeded?~
I see, the difference is in `exec_plan: None,`. I'd suggest to not redefine the whole struct but just do something like:

```
interaction_data.exec_plan = None
```

so the difference is immediate to spot.

File: src/models/batch_auction_model.rs [hunk 4]
```diff
@@ -425,6 +447,59 @@ mod tests {
...
+    }
```
Comment (fedgiac) [inline src/models/batch_auction_model.rs:491]: A nice test would be to also deserialize, should be easy as you could reuse the same values.

**Review by fedgiac — APPROVED** [review]
Looks good!

**Pull request merged**